### PR TITLE
Fireflies

### PR DESCRIPTION
**Pull request by seanharmer**
This branch fixes the fireflies/sparklies and also makes a large improvement to the PBR shaders to get them much closer to reference.

### CODE REVIEW

**Comment by seanharmer** [comment]
09-good-non-metallic-mats

File: src/core/shaders/es3/kuesa_metalrough.inc.frag [hunk 1]
```diff
@@ -54,6 +54,10 @@
 
 #pragma include light.inc.frag
 
+const float M_PI = 3.141592653589793;
```
[the remark this answers is not in the document]
Comment (seanharmer) [reply src/core/shaders/es3/kuesa_metalrough.inc.frag]: Done

File: src/core/shaders/es2/kuesa_metalrough.inc.frag [hunk 2]
```diff
@@ -32,6 +32,10 @@
 
 #pragma include light.inc.frag
 
+const float M_PI = 3.141592653589793;
```
[the remark this answers is not in the document]
Comment (seanharmer) [reply src/core/shaders/es2/kuesa_metalrough.inc.frag]: Done

File: src/core/shaders/es3/kuesa_metalrough.inc.frag [hunk 3]
```diff
@@ -255,7 +246,7 @@ FP vec3 pbrIblModel(const in FP vec3 wNormal,
     // Calculate specular component
     FP vec3 dielectricColor = vec3(0.04);
     FP vec3 F0 = mix(dielectricColor, baseColor, metalness);
-    FP vec3 specularFactor = specularModel(F0, lDotH, lDotN, vDotN, n, h);
+    FP vec3 specularFactor = specularModel(F0, vDotH, lDotN, vDotN, n, h);
```
[the remark this answers is not in the document]
Comment (seanharmer) [reply src/core/shaders/es3/kuesa_metalrough.inc.frag]: Done, hmm not sure how I missed that.

File: src/core/metallicroughnesseffect.h [hunk 4]
```diff
@@ -93,6 +94,7 @@ public Q_SLOTS:
     void setOpaque(bool opaque);
     void setAlphaCutoffEnabled(bool enabled);
     void setToneMappingAlgorithm(ToneMapping algorithm);
+    void setBrdfLUT(Qt3DRender::QAbstractTexture *brdfLUT);
```
[the remark this answers is not in the document]
Comment (seanharmer) [reply src/core/metallicroughnesseffect.h]: Done.

File: src/core/shaders/es2/kuesa_metalrough.inc.frag [hunk 5]
```diff
@@ -114,49 +118,31 @@ FP float alphaToMipLevel(FP float alpha)
...
 }
 
-FP vec3 fresnelFactor(const in FP vec3 color, const in FP float cosineFactor)
+FP vec3 fresnelFactor(const in FP vec3 F0, const in FP vec3 F90, const in FP float cosineFactor)
 {
     // Calculate the Fresnel effect value
-    FP vec3 f = color;
-    FP vec3 F = f + (1.0 - f) * pow(1.0 - cosineFactor, 5.0);
-    return clamp(F, f, vec3(1.0));
+    vec3 F = F0 + (F90 - F0) * pow(clamp(1.0 - cosineFactor, 0.0, 1.0), 5.0);
```
Comment (mkrus) [inline src/core/shaders/es2/kuesa_metalrough.inc.frag]: FP

File: src/core/shaders/es2/kuesa_metalrough.inc.frag [hunk 6]
```diff
@@ -211,28 +196,29 @@ FP vec3 pbrModel(const in int lightIndex,
...
-    FP vec3 color = att * lights[lightIndex].intensity * (specular + diffuse * (vec3(1.0) - specular));
+    // Compute shading terms
+    FP vec3 F = fresnelFactor(F0, F90, vDotH);
+    FP float G = geometricModel(sDotN, vDotN, h, alpha);
+    FP float D = normalDistribution(n, h, alpha);
 
-    // Reduce by ambient occlusion amount
-    color *= ambientOcclusion;
+    // Analytical (punctual) lighting
+    FP vec3 diffuseContrib = (vec3(1.0) - F) * diffuse(diffuseColor);
```
Comment (mkrus) [inline src/core/shaders/es2/kuesa_metalrough.inc.frag]: diffuse() function is missing

File: src/core/shaders/es2/kuesa_metalrough.inc.frag [hunk 7]
```diff
@@ -252,21 +238,19 @@ FP vec3 pbrIblModel(const in FP vec3 wNormal,
...
+    // Calculate diffuse and specular (F0) colors
     FP vec3 dielectricColor = vec3(0.04);
-    FP vec3 F0 = mix(dielectricColor, baseColor, metalness);
-    FP vec3 specularFactor = specularModel(F0, lDotH, lDotN, vDotN, n, h);
+    FP vec3 diffuseColor = baseColor * (vec3(1.0) - dielectricColor);
+    diffuseColor *= (1.0 - metalness);
+    FP vec3 F0 = mix(dielectricColor, baseColor, metalness); // = specularColor
+
+    FP vec2 brdfUV = clamp(vec2(vDotN, 1.0 - sqrt(alpha)), vec2(0.0), vec2(1.0));
+    FP vec2 brdf = texture(brdfLUT, brdfUV).rg;
```
Comment (mkrus) [inline src/core/shaders/es2/kuesa_metalrough.inc.frag]: texture2D()

File: src/core/shaders/es2/kuesa_metalrough.inc.frag (hunk 5 above)
Comment (seanharmer) [reply src/core/shaders/es2/kuesa_metalrough.inc.frag]: Done

File: src/core/shaders/es2/kuesa_metalrough.inc.frag (hunk 6 above)
Comment (seanharmer) [reply src/core/shaders/es2/kuesa_metalrough.inc.frag]: Done

File: src/core/shaders/es2/kuesa_metalrough.inc.frag (hunk 7 above)
Comment (seanharmer) [reply src/core/shaders/es2/kuesa_metalrough.inc.frag]: Done

**Pull request merged**